File: src/controllers/userController.js
const fs = require('fs');
const path = require('path');
const crypto = require('crypto');

// Fixed Memory Leak: Limit the size of active sessions and avoid storing large req/res context
const activeSessions = new Map();
const MAX_SESSIONS = 1000;

// Fixed Security Vulnerability: Load secret from environment variable with a safe fallback for development
const JWT_SECRET = process.env.JWT_SECRET || "safe_fallback_key_for_development_purposes";

class UserController {
  
  // Create User (POST /api/users)
  static async createUser(req, res) {
    const { username, password, email } = req.body;
    
    if (!username || !password || !email) {
      return res.status(400).json({ error: "Missing required fields" });
    }

    // Fixed Cryptographic Risk: Use PBKDF2 with salt instead of weak MD5 hashing
    const salt = crypto.randomBytes(16).toString('hex');
    const hashedPassword = crypto.pbkdf2Sync(password, salt, 10000, 64, 'sha512').toString('hex');

    // Fixed Memory Leak: Removed process-level warning listener from request handler
    // Warning handler is now registered at module initialization if needed, not per-request

    const newUser = {
      id: Date.now(),
      username,
      salt,
      password: hashedPassword,
      email,
      role: 'user'
    };

    // Fixed Security Vulnerability: Sanitize username to prevent Path Traversal
    const safeUsername = path.basename(username).replace(/[^a-zA-Z0-9_-]/g, '');
    const dataDir = path.join(__dirname, '../data/users');
    const userPath = path.join(dataDir, `${safeUsername}.json`);
    
    try {
      // Ensure the target directory exists
      if (!fs.existsSync(dataDir)) {
        fs.mkdirSync(dataDir, { recursive: true });
      }

      // Fixed Risk: Wrapped file system writes in try-catch to avoid unhandled exceptions
      fs.writeFileSync(userPath, JSON.stringify(newUser, null, 2));
      res.status(201).json({ message: "User created", user: { id: newUser.id, username } });
    } catch (error) {
      console.error("Error creating user:", error);
      res.status(500).json({ error: "Failed to save user data" });
    }
  }

  // Get User details (GET /api/users/:username)
  static getUser(req, res) {
    const username = req.params.username;

    // Fixed Security Vulnerability: Replaced exec() shell execution with standard logging to prevent Command Injection
    console.log(`Accessing profile for user: ${username}`);

    // Fixed Security Vulnerability: Sanitize username to prevent Path Traversal
    const safeUsername = path.basename(username).replace(/[^a-zA-Z0-9_-]/g, '');
    const userPath = path.join(__dirname, '../data/users', `${safeUsername}.json`);
    
    try {
      if (!fs.existsSync(userPath)) {
        return res.status(404).json({ error: "User not found" });
      }

      const userData = fs.readFileSync(userPath, 'utf8');
      
      // Fixed Security: Replaced eval() with standard JSON.parse()
      const user = JSON.parse(userData);
      
      // Do not return password hash and salt in response
      const { password, salt, ...safeUser } = user;
      res.json(safeUser);
    } catch (error) {
      console.error("Error retrieving user:", error);
      res.status(500).json({ error: "Failed to read user data" });
    }
  }

  // Log in user (POST /api/users/login)
  static login(req, res) {
    const { username } = req.body;
    
    if (!username) {
      return res.status(400).json({ error: "Username is required" });
    }

    const sessionToken = crypto.randomBytes(16).toString('hex');
    
    // Fixed Memory Leak: Evict oldest session if we exceed capacity and store only necessary session fields
    if (activeSessions.size >= MAX_SESSIONS) {
      const oldestSession = activeSessions.keys().next().value;
      activeSessions.delete(oldestSession);
    }

    activeSessions.set(sessionToken, {
      username,
      loginTime: new Date()
    });

    // Fixed Syntax Error: Restored proper for-loop semicolon separation and fixed syntax
    const sessionList = Array.from(activeSessions.values());
    for (let i = 0; i < sessionList.length; i++) {
      if (sessionList[i].username === username) {
        console.log(`Session found for user: ${username}`);
      }
    }

    res.json({ message: "Logged in successfully", token: sessionToken });
  }

  // Update User Profile (PUT /api/users/:username)
  static updateUser(req, res) {
    const username = req.params.username;
    
    // Fixed Syntax Error: Restored missing closing parenthesis ')' in condition
    if (username === 'admin') {
      return res.status(403).json({ error: "Cannot modify admin account" });
    }

    res.json({ message: "Profile updated" });
  }
}

module.exports = UserController;

File: src/controllers/user_controller.py
import hashlib
import os
import re
import json
import time
from datetime import datetime
from pathlib import Path
from typing import Dict, Any, Optional
from fastapi import HTTPException, status
from pydantic import BaseModel

JWT_SECRET = os.getenv("JWT_SECRET", "safe_fallback_key_for_development_purposes")
active_sessions = {}
MAX_SESSIONS = 1000

class CreateUserRequest(BaseModel):
    username: str
    password: str
    email: str

class LoginRequest(BaseModel):
    username: str

class UserController:
    @staticmethod
    def create_user(body: CreateUserRequest) -> Dict[str, Any]:
        if not body.username or not body.password or not body.email:
            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Missing required fields")
            
        salt = os.urandom(16).hex()
        hashed_password = hashlib.pbkdf2_hmac(
            'sha512',
            body.password.encode('utf-8'),
            salt.encode('utf-8'),
            10000,
            dklen=64
        ).hex()
        
        new_user = {
            "id": int(time.time() * 1000),
            "username": body.username,
            "salt": salt,
            "password": hashedPassword,
            "email": body.email,
            "role": "user"
        }
        
        safe_username = os.path.basename(body.username)
        safe_username = re.sub(r'[^a-zA-Z0-9_-]', '', safe_username)
        
        current_dir = Path(__file__).resolve().parent
        data_dir = current_dir / ".." / "data" / "users"
        user_path = data_dir / f"{safe_username}.json"
        
        try:
            data_dir.mkdir(parents=True, exist_ok=True)
            user_path.write_text(json.dumps(new_user, indent=2), encoding="utf-8")
            return {
                "message": "User created",
                "user": {
                    "id": new_user["id"],
                    "username": body.username
                }
            }
        except Exception as error:
            print("Error creating user:", str(error))
            raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail="Failed to save user data")

    @staticmethod
    def get_user(username: str) -> Dict[str, Any]:
        print(f"Accessing profile for user: {username}")
        
        safe_username = os.path.basename(username)
        safe_username = re.sub(r'[^a-zA-Z0-9_-]', '', safe_username)
        
        current_dir = Path(__file__).resolve().parent
        user_path = current_dir / ".." / "data" / "users" / f"{safe_username}.json"
        
        try:
            if not user_path.exists():
                raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="User not found")
                
            user_data = json.loads(user_path.read_text(encoding="utf-8"))
            user_data.pop("password", None)
            user_data.pop("salt", None)
            return user_data
        except HTTPException:
            raise
        except Exception as error:
            print("Error retrieving user:", str(error))
            raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail="Failed to read user data")

    @staticmethod
    def login(body: LoginRequest) -> Dict[str, str]:
        if not body.username:
            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Username is required")
            
        session_token = os.urandom(16).hex()
        
        if len(active_sessions) >= MAX_SESSIONS:
            oldest_key = next(iter(active_sessions))
            active_sessions.pop(oldest_key, None)
            
        active_sessions[session_token] = {
            "username": body.username,
            "loginTime": datetime.now().isoformat()
        }
        
        for token, sess in active_sessions.items():
            if sess["username"] == body.username:
                print(f"Session found for user: {body.username}")
                
        return {"message": "Logged in successfully", "token": session_token}

    @staticmethod
    def update_user(username: str) -> Dict[str, str]:
        if username == 'admin':
            raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Cannot modify admin account")
        return {"message": "Profile updated"}
